README.md
# Telefonnummer

[![](https://github.com/opendevtools/rescript-telefonnummer/workflows/Release/badge.svg)](https://github.com/opendevtools/rescript-telefonnummer/actions?workflow=Release)
[![npm (scoped)](https://img.shields.io/npm/v/@opendevtools/telefonnummer)](https://npm.im/@opendevtools/telefonnummer)

This is a port of my library [telefonnummer](https://github.com/believer/telefonnummer) to ReasonML. It parses Swedish phone numbers into a standard format. It doesn't contain the full API of the original library as I felt it would be a good opportunity to start fresh.

## Installation

```
npm install @opendevtools/rescript-telefonnummer
```

Add `@opendevtools/rescript-telefonnummer` in `bs-dependencies` of `bsconfig.json`

## API

### parse

Takes a phone number of any kind and parses it to a standard format

```reason
let parse: string => string
```

#### Example

```reason
let parsedPhoneNumber = Telefonnummer.parse("081234567") /* 08-123 45 67 */
let parsedMobileNumber = Telefonnummer.parse("0701234567") /* 070-123 45 67 */
```

### typeOfNumber

Takes a phone number and returns the type. Usable for pattern-matching.

```reason
type t =
  | VoiceMail
  | Mobile
  | Landline

let typeOfNumber: string => t
```

#### Example

```reason
let phoneNumberType = Telefonnummer.typeOfNumber("081234567") /* Landline */
```

### Normalize

The `Normalize` module is exposed publically if you need to clean up a phone number for use in for example `tel:` links.

```reason
module Normalize = {
  let clean: string => string
}
```

#### Example

```reason
let cleanPhoneNumber = Telefonnummer.Normalize.clean("08-12 3 45.67") /* 081234567 */
```

### Validate

The `Validate` module can help you check if a phone number is valid. It cleans the phone number from any non-number characters.

```reason
module Validate = {
  let isValid: string => bool
}
```

#### Example

```reason
let isValid = Telefonnummer.Validate.isValid("08-12 3 45.67") /* true */
```

### Link

The `Link` module creates phone number links.

```reason
module Link = {
  let make: string => string
}
```

#### Example

```reason
let link = Telefonnummer.Link.make("08-12 3 45.67") /* tel:081234567 */
```


CHANGELOG.md
## [2.1.2](https://github.com/opendevtools/rescript-telefonnummer/compare/v2.1.1...v2.1.2) (2021-01-25)


### Bug Fixes

* rename in bsconfig ([7b7e7f5](https://github.com/opendevtools/rescript-telefonnummer/commit/7b7e7f5a23635bb7c25754dff99d2442e8c08f53))

## [2.1.1](https://github.com/opendevtools/rescript-telefonnummer/compare/v2.1.0...v2.1.1) (2020-11-12)


### Bug Fixes

* add publishConfig ([a15b88a](https://github.com/opendevtools/rescript-telefonnummer/commit/a15b88af960bbd82a1c3c7246428315cdd85c612))

# [2.1.0](https://github.com/opendevtools/rescript-telefonnummer/compare/v2.0.0...v2.1.0) (2020-11-12)


### Features

* release under [@opendevtools](https://github.com/opendevtools) organization ([05067dd](https://github.com/opendevtools/rescript-telefonnummer/commit/05067ddd1bad83dd0d56f62b935e482448ce2714))


__tests__/Telefonnummer_test.bs.js
// Generated by BUCKLESCRIPT, PLEASE EDIT WITH CARE
'use strict';

var Jest = require("@glennsl/bs-jest/lib/js/src/jest.js");
var Telefonnummer = require("../src/Telefonnummer.bs.js");

Jest.test("phonenumber link", (function (param) {
        return Jest.Expect.toEqual("tel:0701234567", Jest.Expect.expect(Telefonnummer.Link.make("070-123.45x67")));
      }));

Jest.testAll("voicemail number", {
      hd: "888",
      tl: {
        hd: "333",
        tl: {
          hd: "222",
          tl: {
            hd: "147",
            tl: /* [] */0
          }
        }
      }
    }, (function (phoneNumber) {
        return Jest.Expect.toBe("Röstbrevlåda", Jest.Expect.expect(Telefonnummer.parse(phoneNumber)));
      }));

Jest.testAll("mobile number", {
      hd: [
        "072- 12 34 5 67",
        "072-123 45 67"
      ],
      tl: {
        hd: [
          "0721234567",
          "072-123 45 67"
        ],
        tl: {
          hd: [
            "0701234567",
            "070-123 45 67"
          ],
          tl: {
            hd: [
              "46701234567",
              "070-123 45 67"
            ],
            tl: {
              hd: [
                "+46701234567",
                "070-123 45 67"
              ],
              tl: {
                hd: [
                  "0046701234567",
                  "070-123 45 67"
                ],
                tl: {
                  hd: [
                    "+46(0)701234567",
                    "070-123 45 67"
                  ],
                  tl: {
                    hd: [
                      "+46 0700123456",
                      "070-012 34 56"
                    ],
                    tl: /* [] */0
                  }
                }
              }
            }
          }
        }
      }
    }, (function (param) {
        return Jest.Expect.toBe(param[1], Jest.Expect.expect(Telefonnummer.parse(param[0])));
      }));

Jest.testAll("landline number", {
      hd: [
        "0 8 1 2 34 56",
        "08-12 34 56"
      ],
      tl: {
        hd: [
          "08123456",
          "08-12 34 56"
        ],
        tl: {
          hd: [
            "4608123456",
            "08-12 34 56"
          ],
          tl: {
            hd: [
              "081234567",
              "08-123 45 67"
            ],
            tl: {
              hd: [
                "46081234567",
                "08-123 45 67"
              ],
              tl: {
                hd: [
                  "0812345678",
                  "08-123 456 78"
                ],
                tl: {
                  hd: [
                    "460812345678",
                    "08-123 456 78"
                  ],
                  tl: {
                    hd: [
                      "03112345",
                      "031-123 45"
                    ],
                    tl: {
                      hd: [
                        "031123456",
                        "031-12 34 56"
                      ],
                      tl: {
                        hd: [
                          "004631123456",
                          "031-12 34 56"
                        ],
                        tl: {
                          hd: [
                            "0311234567",
                            "031-123 45 67"
                          ],
                          tl: {
                            hd: [
                              "+460311234567",
                              "031-123 45 67"
                            ],
                            tl: {
                              hd: [
                                "0500123456",
                                "0500-12 34 56"
                              ],
                              tl: {
                                hd: [
                                  "460500123456",
                                  "0500-12 34 56"
                                ],
                                tl: {
                                  hd: [
                                    "05001234567",
                                    "05001234567"
                                  ],
                                  tl: /* [] */0
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }, (function (param) {
        return Jest.Expect.toBe(param[1], Jest.Expect.expect(Telefonnummer.parse(param[0])));
      }));

Jest.testAll("phone number validation", {
      hd: [
        "072- 12 34 5 67",
        true
      ],
      tl: {
        hd: [
          "0721234567",
          true
        ],
        tl: {
          hd: [
            "08123456",
            true
          ],
          tl: {
            hd: [
              "0500123456",
              true
            ],
            tl: {
              hd: [
                "+460311234567",
                true
              ],
              tl: {
                hd: [
                  "05001234567",
                  false
                ],
                tl: {
                  hd: [
                    "a08123456",
                    false
                  ],
                  tl: /* [] */0
                }
              }
            }
          }
        }
      }
    }, (function (param) {
        return Jest.Expect.toBe(param[1], Jest.Expect.expect(Telefonnummer.Validate.isValid(param[0])));
      }));

/*  Not a pure module */


src/Riktnummer.bs.js
// Generated by BUCKLESCRIPT, PLEASE EDIT WITH CARE
'use strict';


var validRiktnummer = [
  [
    "011",
    "Norrk\xc3\xb6ping"
  ],
  [
    "0120",
    "\xc3\x85tvidaberg"
  ],
  [
    "0121",
    "S\xc3\xb6derk\xc3\xb6ping"
  ],
  [
    "0122",
    "Finsp\xc3\xa5ng"
  ],
  [
    "0123",
    "Valdemarsvik"
  ],
  [
    "0125",
    "Vikbolandet"
  ],
  [
    "013",
    "Link\xc3\xb6ping"
  ],
  [
    "0140",
    "Tran\xc3\xa5s"
  ],
  [
    "0141",
    "Motala"
  ],
  [
    "0142",
    "Mj\xc3\xb6lby-Sk\xc3\xa4nninge-Boxholm"
  ],
  [
    "0143",
    "Vadstena"
  ],
  [
    "0144",
    "\xc3\x96desh\xc3\xb6g"
  ],
  [
    "0150",
    "Katrineholm"
  ],
  [
    "0151",
    "Ving\xc3\xa5ker"
  ],
  [
    "0152",
    "Str\xc3\xa4ngn\xc3\xa4s"
  ],
  [
    "0155",
    "Nyk\xc3\xb6ping-Oxel\xc3\xb6sund"
  ],
  [
    "0156",
    "Trosa-Vagnh\xc3\xa4rad"
  ],
  [
    "0157",
    "Flen-Malmk\xc3\xb6ping"
  ],
  [
    "0158",
    "Gnesta"
  ],
  [
    "0159",
    "Mariefred"
  ],
  [
    "016",
    "Eskilstuna-Torsh\xc3\xa4lla"
  ],
  [
    "0171",
    "Enk\xc3\xb6ping"
  ],
  [
    "0173",
    "\xc3\x96regrund-\xc3\x96sthammar"
  ],
  [
    "0174",
    "Alunda"
  ],
  [
    "0175",
    "Hallstavik-Rimbo"
  ],
  [
    "0176",
    "Norrt\xc3\xa4lje"
  ],
  [
    "018",
    "Uppsala"
  ],
  [
    "019",
    "\xc3\x96rebro-Kumla"
  ],
  [
    "021",
    "V\xc3\xa4ster\xc3\xa5s"
  ],
  [
    "0220",
    "Hallstahammar-Surahammar"
  ],
  [
    "0221",
    "K\xc3\xb6ping"
  ],
  [
    "0222",
    "Skinnskatteberg"
  ],
  [
    "0223",
    "Fagersta-Norberg"
  ],
  [
    "0224",
    "Sala-Heby"
  ],
  [
    "0225",
    "Hedemora-S\xc3\xa4ter"
  ],
  [
    "0226",
    "Avesta-Krylbo"
  ],
  [
    "0227",
    "Kungs\xc3\xb6r"
  ],
  [
    "023",
    "Falun"
  ],
  [
    "0240",
    "Ludvika-Smedjebacken"
  ],
  [
    "0241",
    "Gagnef-Floda"
  ],
  [
    "0243",
    "Borl\xc3\xa4nge"
  ],
  [
    "0246",
    "Sv\xc3\xa4rdsj\xc3\xb6-Enviken"
  ],
  [
    "0247",
    "Leksand-Insj\xc3\xb6n"
  ],
  [
    "0248",
    "R\xc3\xa4ttvik"
  ],
  [
    "0250",
    "Mora-Orsa"
  ],
  [
    "0251",
    "\xc3\x84lvdalen"
  ],
  [
    "0253",
    "Idre-S\xc3\xa4rna"
  ],
  [
    "0258",
    "Furudal"
  ],
  [
    "026",
    "G\xc3\xa4vle-Sandviken"
  ],
  [
    "0270",
    "S\xc3\xb6derhamn"
  ],
  [
    "0271",
    "Alfta-Edsbyn"
  ],
  [
    "0278",
    "Bolln\xc3\xa4s"
  ],
  [
    "0280",
    "Malung"
  ],
  [
    "0281",
    "Vansbro"
  ],
  [
    "0290",
    "Hofors-Storvik"
  ],
  [
    "0291",
    "Hedesunda-\xc3\x96sterf\xc3\xa4rnebo"
  ],
  [
    "0292",
    "T\xc3\xa4rnsj\xc3\xb6-\xc3\x96sterv\xc3\xa5la"
  ],
  [
    "0293",
    "Tierp-S\xc3\xb6derfors"
  ],
  [
    "0294",
    "Karlholmsbruk-Sk\xc3\xa4rplinge"
  ],
  [
    "0295",
    "\xc3\x96rbyhus-Dannemora"
  ],
  [
    "0297",
    "Ockelbo-Hamr\xc3\xa5nge"
  ],
  [
    "0300",
    "Kungsbacka"
  ],
  [
    "0301",
    "Hind\xc3\xa5s"
  ],
  [
    "0302",
    "Lerum"
  ],
  [
    "0303",
    "Kung\xc3\xa4lv"
  ],
  [
    "0304",
    "Orust-Tj\xc3\xb6rn"
  ],
  [
    "031",
    "G\xc3\xb6teborg"
  ],
  [
    "0320",
    "Kinna"
  ],
  [
    "0321",
    "Ulricehamn"
  ],
  [
    "0322",
    "Alings\xc3\xa5s-V\xc3\xa5rg\xc3\xa5rda"
  ],
  [
    "0325",
    "Svenljunga-Tranemo"
  ],
  [
    "033",
    "Bor\xc3\xa5s"
  ],
  [
    "0340",
    "Varberg"
  ],
  [
    "0345",
    "Hyltebruk-Torup"
  ],
  [
    "0346",
    "Falkenberg"
  ],
  [
    "035",
    "Halmstad"
  ],
  [
    "036",
    "J\xc3\xb6nk\xc3\xb6ping-Huskvarna"
  ],
  [
    "0370",
    "V\xc3\xa4rnamo"
  ],
  [
    "0371",
    "Gislaved-Anderstorp"
  ],
  [
    "0372",
    "Ljungby"
  ],
  [
    "0380",
    "N\xc3\xa4ssj\xc3\xb6"
  ],
  [
    "0381",
    "Eksj\xc3\xb6"
  ],
  [
    "0382",
    "S\xc3\xa4vsj\xc3\xb6"
  ],
  [
    "0383",
    "Vetlanda"
  ],
  [
    "0390",
    "Gr\xc3\xa4nna"
  ],
  [
    "0392",
    "Mullsj\xc3\xb6"
  ],
  [
    "0393",
    "Vaggeryd"
  ],
  [
    "040",
    "Malm\xc3\xb6"
  ],
  [
    "0410",
    "Trelleborg"
  ],
  [
    "0411",
    "Ystad"
  ],
  [
    "0413",
    "Esl\xc3\xb6v-H\xc3\xb6\xc3\xb6r"
  ],
  [
    "0414",
    "Simrishamn"
  ],
  [
    "0415",
    "H\xc3\xb6rby"
  ],
  [
    "0416",
    "Sj\xc3\xb6bo"
  ],
  [
    "0417",
    "Tomelilla"
  ],
  [
    "0418",
    "Landskrona-Sval\xc3\xb6v"
  ],
  [
    "042",
    "Helsingborg-H\xc3\xb6gan\xc3\xa4s"
  ],
  [
    "0430",
    "Laholm"
  ],
  [
    "0431",
    "\xc3\x84ngelholm-B\xc3\xa5stad"
  ],
  [
    "0433",
    "Markaryd-Str\xc3\xb6msn\xc3\xa4sbruk"
  ],
  [
    "0435",
    "Klippan-Perstorp"
  ],
  [
    "044",
    "Kristianstad"
  ],
  [
    "0451",
    "H\xc3\xa4ssleholm"
  ],
  [
    "0454",
    "Karlshamn-Olofstr\xc3\xb6m"
  ],
  [
    "0455",
    "Karlskrona"
  ],
  [
    "0456",
    "S\xc3\xb6lvesborg-Brom\xc3\xb6lla"
  ],
  [
    "0457",
    "Ronneby"
  ],
  [
    "0459",
    "Ryd"
  ],
  [
    "046",
    "Lund"
  ],
  [
    "0470",
    "V\xc3\xa4xj\xc3\xb6"
  ],
  [
    "0471",
    "Emmaboda"
  ],
  [
    "0472",
    "Alvesta-Rydaholm"
  ],
  [
    "0474",
    "\xc3\x85seda-Lenhovda"
  ],
  [
    "0476",
    "\xc3\x84lmhult"
  ],
  [
    "0477",
    "Tingsryd"
  ],
  [
    "0478",
    "Lessebo"
  ],
  [
    "0479",
    "Osby"
  ],
  [
    "0480",
    "Kalmar"
  ],
  [
    "0481",
    "Nybro"
  ],
  [
    "0485",
    "\xc3\x96land"
  ],
  [
    "0486",
    "Tors\xc3\xa5s"
  ],
  [
    "0490",
    "V\xc3\xa4stervik"
  ],
  [
    "0491",
    "Oskarshamn-H\xc3\xb6gsby"
  ],
  [
    "0492",
    "Vimmerby"
  ],
  [
    "0493",
    "Gamleby"
  ],
  [
    "0494",
    "Kisa"
  ],
  [
    "0495",
    "Hultsfred-Virserum"
  ],
  [
    "0496",
    "Mariannelund"
  ],
  [
    "0498",
    "Gotland"
  ],
  [
    "0499",
    "M\xc3\xb6nster\xc3\xa5s"
  ],
  [
    "0500",
    "Sk\xc3\xb6vde"
  ],
  [
    "0501",
    "Mariestad"
  ],
  [
    "0502",
    "Tidaholm"
  ],
  [
    "0503",
    "Hjo"
  ],
  [
    "0504",
    "Tibro"
  ],
  [
    "0505",
    "Karlsborg"
  ],
  [
    "0506",
    "T\xc3\xb6reboda-Hova"
  ],
  [
    "0510",
    "Lidk\xc3\xb6ping"
  ],
  [
    "0511",
    "Skara-G\xc3\xb6tene"
  ],
  [
    "0512",
    "Vara-Nossebro"
  ],
  [
    "0513",
    "Herrljunga"
  ],
  [
    "0514",
    "Gr\xc3\xa4storp"
  ],
  [
    "0515",
    "Falk\xc3\xb6ping"
  ],
  [
    "0520",
    "Trollh\xc3\xa4ttan"
  ],
  [
    "0521",
    "V\xc3\xa4nersborg"
  ],
  [
    "0522",
    "Uddevalla"
  ],
  [
    "0523",
    "Lysekil"
  ],
  [
    "0524",
    "Munkedal"
  ],
  [
    "0525",
    "Grebbestad"
  ],
  [
    "0526",
    "Str\xc3\xb6mstad"
  ],
  [
    "0528",
    "F\xc3\xa4rgelanda"
  ],
  [
    "0530",
    "Mellerud"
  ],
  [
    "0531",
    "Bengtsfors"
  ],
  [
    "0532",
    "\xc3\x85m\xc3\xa5l"
  ],
  [
    "0533",
    "S\xc3\xa4ffle"
  ],
  [
    "0534",
    "Ed"
  ],
  [
    "054",
    "Karlstad"
  ],
  [
    "0550",
    "Kristinehamn"
  ],
  [
    "0551",
    "Gullsp\xc3\xa5ng"
  ],
  [
    "0552",
    "Deje"
  ],
  [
    "0553",
    "Molkom"
  ],
  [
    "0554",
    "Kil"
  ],
  [
    "0555",
    "Grums"
  ],
  [
    "0560",
    "Torsby"
  ],
  [
    "0563",
    "Hagfors-Munkfors"
  ],
  [
    "0564",
    "Syssleb\xc3\xa4ck"
  ],
  [
    "0565",
    "Sunne"
  ],
  [
    "0570",
    "Arvika"
  ],
  [
    "0571",
    "Charlottenberg-\xc3\x85motfors"
  ],
  [
    "0573",
    "\xc3\x85rj\xc3\xa4ng"
  ],
  [
    "0580",
    "Kopparberg"
  ],
  [
    "0581",
    "Lindesberg"
  ],
  [
    "0582",
    "Hallsberg"
  ],
  [
    "0583",
    "Askersund"
  ],
  [
    "0584",
    "Lax\xc3\xa5"
  ],
  [
    "0585",
    "Fjugesta-Svart\xc3\xa5"
  ],
  [
    "0586",
    "Karlskoga-Degerfors"
  ],
  [
    "0587",
    "Nora"
  ],
  [
    "0589",
    "Arboga"
  ],
  [
    "0590",
    "Filipstad"
  ],
  [
    "0591",
    "H\xc3\xa4llefors-Grythyttan"
  ],
  [
    "060",
    "Sundsvall-Timr\xc3\xa5"
  ],
  [
    "0611",
    "H\xc3\xa4rn\xc3\xb6sand"
  ],
  [
    "0612",
    "Kramfors"
  ],
  [
    "0613",
    "Ull\xc3\xa5nger"
  ],
  [
    "0620",
    "Sollefte\xc3\xa5"
  ],
  [
    "0621",
    "Junsele"
  ],
  [
    "0622",
    "N\xc3\xa4s\xc3\xa5ker"
  ],
  [
    "0623",
    "Ramsele"
  ],
  [
    "0624",
    "Backe"
  ],
  [
    "063",
    "\xc3\x96stersund"
  ],
  [
    "0640",
    "Krokom"
  ],
  [
    "0642",
    "Lit"
  ],
  [
    "0643",
    "Hallen-Oviken"
  ],
  [
    "0644",
    "Hammerdal"
  ],
  [
    "0645",
    "F\xc3\xb6llinge"
  ],
  [
    "0647",
    "\xc3\x85re-J\xc3\xa4rpen"
  ],
  [
    "0650",
    "Hudiksvall"
  ],
  [
    "0651",
    "Ljusdal"
  ],
  [
    "0652",
    "Bergsj\xc3\xb6"
  ],
  [
    "0653",
    "Delsbo"
  ],
  [
    "0657",
    "Los"
  ],
  [
    "0660",
    "\xc3\x96rnsk\xc3\xb6ldsvik"
  ],
  [
    "0661",
    "Bredbyn"
  ],
  [
    "0662",
    "Bj\xc3\xb6rna"
  ],
  [
    "0663",
    "Husum"
  ],
  [
    "0670",
    "Str\xc3\xb6msund"
  ],
  [
    "0671",
    "Hoting"
  ],
  [
    "0672",
    "G\xc3\xa4ddede"
  ],
  [
    "0680",
    "Sveg"
  ],
  [
    "0682",
    "R\xc3\xa4tan"
  ],
  [
    "0684",
    "Hede-Fun\xc3\xa4sdalen"
  ],
  [
    "0687",
    "Svenstavik"
  ],
  [
    "0690",
    "\xc3\x85nge"
  ],
  [
    "0691",
    "Torpshammar"
  ],
  [
    "0692",
    "Liden"
  ],
  [
    "0693",
    "Br\xc3\xa4cke-G\xc3\xa4ll\xc3\xb6"
  ],
  [
    "0695",
    "Stugun"
  ],
  [
    "0696",
    "Hammarstrand"
  ],
  [
    "08",
    "Stockholm"
  ],
  [
    "090",
    "Ume\xc3\xa5"
  ],
  [
    "0910",
    "Skellefte\xc3\xa5"
  ],
  [
    "0911",
    "Pite\xc3\xa5"
  ],
  [
    "0912",
    "Byske"
  ],
  [
    "0913",
    "L\xc3\xb6v\xc3\xa5nger"
  ],
  [
    "0914",
    "Burtr\xc3\xa4sk"
  ],
  [
    "0915",
    "Bastutr\xc3\xa4sk"
  ],
  [
    "0916",
    "J\xc3\xb6rn"
  ],
  [
    "0918",
    "Norsj\xc3\xb6"
  ],
  [
    "0920",
    "Lule\xc3\xa5"
  ],
  [
    "0921",
    "Boden"
  ],
  [
    "0922",
    "Haparanda"
  ],
  [
    "0923",
    "Kalix"
  ],
  [
    "0924",
    "R\xc3\xa5ne\xc3\xa5"
  ],
  [
    "0925",
    "Lakatr\xc3\xa4sk"
  ],
  [
    "0926",
    "\xc3\x96verkalix"
  ],
  [
    "0927",
    "\xc3\x96vertorne\xc3\xa5"
  ],
  [
    "0928",
    "Harads"
  ],
  [
    "0929",
    "\xc3\x84lvsbyn"
  ],
  [
    "0930",
    "Nordmaling"
  ],
  [
    "0932",
    "Bjurholm"
  ],
  [
    "0933",
    "Vindeln"
  ],
  [
    "0934",
    "Robertsfors"
  ],
  [
    "0935",
    "V\xc3\xa4nn\xc3\xa4s"
  ],
  [
    "0940",
    "Vilhelmina"
  ],
  [
    "0941",
    "\xc3\x85sele"
  ],
  [
    "0942",
    "Dorotea"
  ],
  [
    "0943",
    "Fredrika"
  ],
  [
    "0950",
    "Lycksele"
  ],
  [
    "0951",
    "Storuman"
  ],
  [
    "0952",
    "Sorsele"
  ],
  [
    "0953",
    "Mal\xc3\xa5"
  ],
  [
    "0954",
    "T\xc3\xa4rnaby"
  ],
  [
    "0960",
    "Arvidsjaur"
  ],
  [
    "0961",
    "Arjeplog"
  ],
  [
    "0970",
    "G\xc3\xa4llivare"
  ],
  [
    "0971",
    "Jokkmokk"
  ],
  [
    "0973",
    "Porjus"
  ],
  [
    "0975",
    "Hakkas"
  ],
  [
    "0976",
    "Vuollerim"
  ],
  [
    "0977",
    "Korpilombolo"
  ],
  [
    "0978",
    "Pajala"
  ],
  [
    "0980",
    "Kiruna"
  ],
  [
    "0981",
    "Vittangi"
  ]
];

exports.validRiktnummer = validRiktnummer;
/* No side effect */


src/Telefonnummer.bs.js
// Generated by BUCKLESCRIPT, PLEASE EDIT WITH CARE
'use strict';

var Curry = require("bs-platform/lib/js/curry.js");
var Riktnummer = require("./Riktnummer.bs.js");

function clean(phoneNumber) {
  var normalized = phoneNumber.replace("(0)", "").replace(/\D+/gi, "");
  var withPhone = function (from) {
    return "0" + normalized.substr(from);
  };
  var pn = normalized.substring(0, 4);
  if (pn.startsWith("0046")) {
    return withPhone(4);
  } else if (pn.startsWith("460")) {
    return withPhone(3);
  } else if (pn.startsWith("46")) {
    return withPhone(2);
  } else {
    return normalized;
  }
}

var Normalize = {
  clean: clean
};

function make(phoneNumber) {
  return "tel:" + clean(phoneNumber);
}

var Link = {
  make: make
};

function replacer(regex, replaceWithOpt, param) {
  var replaceWith = replaceWithOpt !== undefined ? replaceWithOpt : "$1-$2 $3 $4";
  return function (param) {
    return param.replace(regex, replaceWith);
  };
}

function fiveDigit(areaCode) {
  return new RegExp("^(\\d{" + String(areaCode) + "})(\\d{3})(\\d{2})$");
}

function sixDigit(areaCode) {
  return new RegExp("^(\\d{" + String(areaCode) + "})(\\d{2})(\\d{2})(\\d{2})$");
}

function sevenDigit(areaCode) {
  return new RegExp("^(\\d{" + String(areaCode) + "})(\\d{3})(\\d{2})(\\d{2})$");
}

function eightDigit(areaCode) {
  return new RegExp("^(\\d{" + String(areaCode) + "})(\\d{3})(\\d{3})(\\d{2})$");
}

var regex = /^08/;

var sixDigit$1 = replacer(sixDigit(2), undefined, undefined);

var sevenDigit$1 = replacer(sevenDigit(2), undefined, undefined);

var eightDigit$1 = replacer(eightDigit(2), undefined, undefined);

var regex$1 = /^0(1[013689]|2[0136]|3[1356]|4[0246]|54|6[03]|7[0235-9]|9[09])/;

var fiveDigit$1 = replacer(fiveDigit(3), "$1-$2 $3", undefined);

var sixDigit$2 = replacer(sixDigit(3), undefined, undefined);

var sevenDigit$2 = replacer(sevenDigit(3), undefined, undefined);

var sixDigit$3 = replacer(sixDigit(4), undefined, undefined);

function digits(value) {
  if (regex.test(value)) {
    return "Two";
  } else if (regex$1.test(value)) {
    return "Three";
  } else {
    return "Four";
  }
}

var phoneNumbers = [
  "888",
  "333",
  "222",
  "147"
];

var valid = /^07(0|2|3|6|9)\d{7}$/;

function findValidByRiktnummer(digits, trailingDigits) {
  var codes = Riktnummer.validRiktnummer.filter(function (param) {
            return param[0].length === digits;
          }).map(function (param) {
          return param[0];
        }).join("|");
  return new RegExp("^(" + codes + ")\\d{5," + String(trailingDigits) + "}$");
}

var validTwoDigit = /^08\d{6,7}$/;

var validThreeDigit = findValidByRiktnummer(3, 7);

var validFourDigit = findValidByRiktnummer(4, 6);

function typeOfNumber(number) {
  if (number.normalize().startsWith("07")) {
    return /* Mobile */1;
  } else if (phoneNumbers.includes(number)) {
    return /* VoiceMail */0;
  } else {
    return /* Landline */2;
  }
}

function parse(phoneNumber) {
  var match = typeOfNumber(phoneNumber);
  switch (match) {
    case /* VoiceMail */0 :
        return "Röstbrevlåda";
    case /* Mobile */1 :
        return Curry._1(sevenDigit$2, clean(phoneNumber));
    case /* Landline */2 :
        var pn = clean(phoneNumber);
        var match$1 = digits(pn);
        var match$2 = pn.length;
        var tmp;
        var exit = 0;
        if (match$1 === "Two") {
          switch (match$2) {
            case 8 :
                tmp = sixDigit$1;
                break;
            case 9 :
                tmp = sevenDigit$1;
                break;
            case 10 :
                tmp = eightDigit$1;
                break;
            default:
              exit = 1;
          }
        } else if (match$1 === "Three") {
          switch (match$2) {
            case 8 :
                tmp = fiveDigit$1;
                break;
            case 9 :
                tmp = sixDigit$2;
                break;
            case 10 :
                tmp = sevenDigit$2;
                break;
            default:
              exit = 1;
          }
        } else if (match$1 === "Four") {
          tmp = sixDigit$3;
        } else {
          exit = 1;
        }
        if (exit === 1) {
          tmp = (function (param) {
              return pn;
            });
        }
        return tmp(pn);
    
  }
}

function isValid(phoneNumber) {
  if (/[a-z]/gi.test(phoneNumber)) {
    return false;
  }
  var normalized = clean(phoneNumber);
  var digits$1 = digits(normalized);
  var typeOfNumber$1 = typeOfNumber(normalized);
  switch (typeOfNumber$1) {
    case /* VoiceMail */0 :
        return true;
    case /* Mobile */1 :
        return valid.test(normalized);
    case /* Landline */2 :
        if (digits$1 === "Three") {
          return validThreeDigit.test(normalized);
        } else if (digits$1 === "Four") {
          return validFourDigit.test(normalized);
        } else {
          return validTwoDigit.test(normalized);
        }
    
  }
}

var Validate = {
  isValid: isValid
};

exports.Validate = Validate;
exports.Normalize = Normalize;
exports.Link = Link;
exports.typeOfNumber = typeOfNumber;
exports.parse = parse;
/* sixDigit Not a pure module */
